python/lsst/cp/pipe/utils.py

```diff
@@ -764,3 +765,262 @@ def ddict2dict(d):
         if isinstance(v, dict):
             d[k] = ddict2dict(v)
     return dict(d)
+
+
+class AstierSplineLinearityFitter:
+    """Class to fit the Astier spline linearity model.
+
+    This is a spline fit with photodiode data based on a model
+    from Pierre Astier.
+    This model fits a spline with (optional) nuisance parameters
+    to allow for different linearity coefficients with different
+    photodiode settings.  The minimization is a least-squares
+    fit with the residual of
+    Sum[(S(mu_i) + mu_i)/(k_j * D_i) - 1]**2, where mu_i is the
+    observed flat-pair mean; D_j is the photo-diode measurement
+    corresponding to that flat-pair; and k_j is a constant of
+    proportionality which is over index j as it is allowed to
+    be different based on different photodiode settings (e.g.
+    CCOBCURR).
+
+    The fit has additional constraints to ensure that the spline
+    goes through the (0, 0) point, as well as a normalization
+    condition so that the average of the spline over the full
```

Comment:
Can you explain this normalization condition more?  Is this trying to conserve charge?  I ask because if we attempted to spline fit something with a simple squared correction, it seems like this would cause problems.

Reply:
This is from Pierre's code, and what I think it's basically doing is what I said: the total sum of the spline term should be zero, so that the linear part is picked up by the linear fit.  I think this is basically the same thing that happens in the squared correction as well, as it's a squared deviation from linear, but the linear term is dropped in the application.

Reply:
I've added this comment:
```
    The fit has additional constraints to ensure that the spline
    goes through the (0, 0) point, as well as a normalization
    condition so that the average of the spline over the full
    range is 0. The normalization ensures that the spline only
    fits deviations from linearity, rather than the linear
    function itself which is degenerate with the gain.
```